### PR TITLE
EBODAC-668: Added Screen for Prime Vaccination

### PR DESCRIPTION
**Pull request by ngraczewski**

### CODE REVIEW

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    private Long visitId;
+
+    public VisitBookingDetailsDto() {
+    }
+
+    public VisitBookingDetailsDto(LocalDate date, Subject participant) {
+        participantId = participant.getSubjectId();
+        participantName = participant.getName() + " " + participant.getHouseholdName();
+        femaleChildBearingAge = participant.getGender() == Gender.Female;
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java]: Seth requested this not to be calculated but selectable on the UI

File: ebodac/src/main/java/org/motechproject/ebodac/repository/VisitDataService.java [hunk 2]
```diff
@@ -36,6 +37,11 @@
     @Lookup
     List<Visit> findVisitsByType(@LookupField(name = "type") VisitType visitType);
 
+    @Lookup
+    List<Visit> findVisitsByTypeAndDate(@LookupField(name = "type") VisitType visitType,
```
Comment (pmuchowski) [inline ebodac/src/main/java/org/motechproject/ebodac/repository/VisitDataService.java]: I think that you need to check if Prime Vaccination Date in the Participant data is null

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java [hunk 3]
```diff
@@ -43,6 +45,36 @@ public VisitBookingDetails(Visit visit, LocalDate bookingActualDate) {
         this.bookingActualDate = bookingActualDate;
     }
 
+    public VisitBookingDetailsDto toDto() {
+
+        Subject participant = getVisit().getSubject();
+
+        VisitBookingDetailsDto dto = new VisitBookingDetailsDto();
+        dto.setBookingId(getId());
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java]: Id of the Visit Detail is not the Booking Id, if you need this rename it

File: booking-app/src/main/resources/webapp/js/directives.js [hunk 4]
```diff
@@ -107,4 +162,104 @@
...
+                        scope.msg("bookingApp.primeVaccination.location"),
+                        scope.msg("bookingApp.primeVaccination.participantId"),
+                        scope.msg("bookingApp.primeVaccination.participantName"),
+                        scope.msg("bookingApp.primeVaccination.femaleChildBearingAge"),
+                        scope.msg("bookingApp.primeVaccination.screeningActualDate"),
+                        scope.msg("bookingApp.date"),
+                        scope.msg("bookingApp.startTime"),
+                        scope.msg("bookingApp.endTime"), ""],
+                    colModel: [
+                        { name: "bookingId"},
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/directives.js]: Why are you using this Id here?

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 5]
```diff
@@ -3,10 +3,94 @@
...
+                winPrint.focus();
+                winPrint.print();
+            }
+        }
+
+        $scope.reloadSelects = function() {
+            $timeout(function() {
+                $('#siteSelect').trigger('change');
+                $('#clinicSelect').trigger('change');
+                $('#roomSelect').trigger('change');
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js]: room is not used any more

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 6]
```diff
@@ -3,10 +3,94 @@
 
     var controllers = angular.module('bookingApp.controllers', []);
 
-    controllers.controller('BAScreeningCtrl', function ($scope, $timeout, $http, Screenings, Sites) {
+    controllers.controller('BABaseCtrl', function ($scope, $timeout, Screenings, Sites) {
 
         $scope.sites = Sites.query();
         $scope.screeningForPrint = {};
+
+        $scope.printRow = function(id) {
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js]: why this is in the Base controller?

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java [hunk 7]
```diff
@@ -43,6 +45,36 @@ public VisitBookingDetails(Visit visit, LocalDate bookingActualDate) {
...
+        dto.setParticipantId(participant.getSubjectId());
+        dto.setParticipantName(participant.getName() + " " + participant.getHouseholdName());
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java]: why are you adding Household Name to Participant Name?

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+import org.codehaus.jackson.map.annotate.JsonSerialize;
+import org.joda.time.LocalDate;
+import org.motechproject.bookingapp.util.CustomDateDeserializer;
+import org.motechproject.bookingapp.util.CustomDateSerializer;
+import org.motechproject.bookingapp.util.CustomTimeSerializer;
+import org.motechproject.commons.date.model.Time;
+import org.motechproject.ebodac.domain.Gender;
+import org.motechproject.ebodac.domain.Subject;
+
+public class VisitBookingDetailsDto {
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java]: I don't think this should be named VisitBookingDetailsDto is more something like PrimeVaccinationScheduleDto to me

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java [hunk 9]
```diff
@@ -43,6 +45,36 @@ public VisitBookingDetails(Visit visit, LocalDate bookingActualDate) {
...
+        dto.setFemaleChildBearingAge(isFemaleChildBearingAge(participant));
+
+        for (Visit participantVisit : participant.getVisits()) {
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java]: if you using this only with Screening details you don't need to iterate through all visits, just use getVisit() instead

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java [hunk 10]
```diff
@@ -43,6 +45,36 @@ public VisitBookingDetails(Visit visit, LocalDate bookingActualDate) {
         this.bookingActualDate = bookingActualDate;
     }
 
+    public VisitBookingDetailsDto toDto() {
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java]: I don't think this logic should be added here, because this Dto should be something like PrimeVaccinationScheduleDto

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java [hunk 11]
```diff
@@ -43,6 +45,36 @@ public VisitBookingDetails(Visit visit, LocalDate bookingActualDate) {
...
+        dto.setStartTime(getStartTime());
+        dto.setEndTime(getEndTime());
+        dto.setClinicId(getClinic().getId());
+        dto.setSiteId(getClinic().getSite().getId());
+        dto.setVisitId(getVisit().getId());
+        dto.setLocation(getClinic().getSite().getSiteId() + " - " + getClinic().getLocation());
+        return dto;
+    }
+
+    private boolean isFemaleChildBearingAge(Subject participant) {
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetails.java]: Seth requested that this should not be calculated, but user should choose on the UI the value

File: booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java (hunk 8 above)
Comment (kkaczmarczyk) [inline booking-app/src/main/java/org/motechproject/bookingapp/domain/VisitBookingDetailsDto.java]: agreed - this dto is more related to the Prime Vac screen then simply to VisitBookingDetails

File: booking-app/src/main/java/org/motechproject/bookingapp/service/impl/VisitBookingDetailsServiceImpl.java [hunk 12]
```diff
@@ -74,4 +88,120 @@ public void createOrUpdate(Collection<VisitBookingDetails> visitBookingDetailsLi
             }
         }
     }
+
+    @Override
+    public List<VisitBookingDetailsDto> getVisitDtos(int page, int pageSize, String sortColumn, String sortDirection) {
```
Comment (kkaczmarczyk) [inline booking-app/src/main/java/org/motechproject/bookingapp/service/impl/VisitBookingDetailsServiceImpl.java]: let's move this logic to a new servie called PrimeVaccinationScheduleService - this isn't striclty related to VisitBookingDetails

File: booking-app/src/main/java/org/motechproject/bookingapp/service/VisitBookingDetailsService.java [hunk 13]
```diff
@@ -23,4 +24,12 @@
     VisitBookingDetails createOrUpdate(VisitBookingDetails visitBookingDetails);
 
     void createOrUpdate(Collection<VisitBookingDetails> visitBookingDetailsList);
+
+    List<VisitBookingDetailsDto> getVisitDtos(int page, int pageSize, String sortColumn, String sortDirection);
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/service/VisitBookingDetailsService.java]: I think this logic should be move to separate service PrimeVaccinationScheduleService, @kkaczmarczyk what do you think?

File: booking-app/src/main/java/org/motechproject/bookingapp/web/VisitBookingDetailsController.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import java.util.List;
+
+@Controller
+public class VisitBookingDetailsController {
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/web/VisitBookingDetailsController.java]: rename it to PrimeVaccinationScheduleController

File: booking-app/src/main/resources/webapp/js/app.js [hunk 15]
```diff
@@ -21,7 +21,7 @@
             })
             .when('/bookingApp/primeVaccination', {
                 templateUrl: '../booking-app/resources/partials/primeVaccination.html',
-                controller: 'BookingAppPrimeVaccinationCtrl'
+                controller: 'BAPrimeVaccinationCtrl'
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/app.js]: I think we should use one naming in all controllers, for me BookingApp looks better than BA, @kkaczmarczyk what do you think?

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 16]
```diff
@@ -92,76 +157,53 @@
...
-
-            if (startTime.hours === endTime.hours) {
-                return startTime.minutes < endTime.minutes;
-            }
-
-            return startTime.hours < endTime.hours;
-        }
-
-        $scope.printRow = function(id) {
+        $scope.addScreening = function() {
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js]: why this is named addScreening?

File: booking-app/src/main/resources/webapp/js/services.js [hunk 17]
```diff
@@ -5,20 +5,26 @@
...
 
     services.factory('Sites', function($resource) {
-        return $resource('../booking-app/sites', {}, {});
+        return $resource('../booking-app/sites', {}, {
+            'getClinics': { url: '../booking-app/sites/:id/clinics', method: 'GET', isArray: true },
+            'getRooms': { url: '../booking-app/sites/clinics/:id/rooms', method: 'GET', isArray: true }
+        });
+    });
+
+    services.factory('VisitBookingDetails', function($resource) {
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/services.js]: change the name to PrimeVaccinationSchedule or something similar

File: booking-app/src/main/resources/webapp/messages/messages.properties [hunk 18]
```diff
@@ -5,29 +5,39 @@ bookingApp.primeVaccination=Prime Vaccination
...
-bookingApp.screening.endTimePlaceholder=Click to select end time
 bookingApp.screening.from=From
 bookingApp.screening.name=Name
 bookingApp.screening.namePlaceholder=Enter name
 bookingApp.screening.printCard=Print card
-bookingApp.screening.save=Save
 bookingApp.screening.scheduleScreening=Schedule screening
 bookingApp.screening.scheduleSuccessful=Screening visit scheduled successfully.
-bookingApp.screening.error=Error
-bookingApp.screening.scheduleError=Error occurred during scheduling Screening Visit\: {0}
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/messages/messages.properties:27]: why did you remove this?

File: booking-app/src/main/resources/webapp/partials/modals/primeVaccinationModal.html [hunk 19]
```diff
@@ -0,0 +1,74 @@
+<div class="modal fade" id="visitBookingDetailsModal" tabindex="-1" role="dialog">
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/partials/modals/primeVaccinationModal.html]: change the id

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 20]
```diff
@@ -92,76 +157,53 @@
...
+                $scope.reloadSelects();
+            });
+        };
 
-            var winPrint = window.open("../booking-app/resources/partials/volunteerCardScreening.html");
-            winPrint.onload = function() {
-                $('#bookingId', winPrint.document).html(bookingId);
-                $('#volunteerName', winPrint.document).html(volunteerName);
-                $('#screeningDate', winPrint.document).html(date);
+        $scope.saveVisitBookingDetails = function() {
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js]: change the name of this

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 21]
```diff
@@ -3,10 +3,94 @@
...
+
+        $scope.reloadSelects = function() {
+            $timeout(function() {
+                $('#siteSelect').trigger('change');
+                $('#clinicSelect').trigger('change');
+                $('#roomSelect').trigger('change');
+            });
+        };
+
+        $scope.findById = function(list, id) {
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js:18]: why this is in the basic controller?

File: booking-app/src/main/java/org/motechproject/bookingapp/service/impl/VisitBookingDetailsServiceImpl.java [hunk 22]
```diff
@@ -74,4 +88,120 @@ public void createOrUpdate(Collection<VisitBookingDetails> visitBookingDetailsLi
...
+
+        QueryParams queryParams = buildQueryParams(page, pageSize, sortColumn, sortDirection);
+
+        List<Visit> visits = visitDataService.findVisitsByTypeAndDate(VisitType.SCREENING, null, queryParams);
```
Comment (pmuchowski) [inline booking-app/src/main/java/org/motechproject/bookingapp/service/impl/VisitBookingDetailsServiceImpl.java]: you need to check if Participant has no prime Vaccination date also

File: booking-app/src/main/resources/webapp/js/app.js (hunk 15 above)
Comment (kkaczmarczyk) [inline booking-app/src/main/resources/webapp/js/app.js]: Agreed

File: booking-app/src/main/resources/webapp/js/controllers.js [hunk 23]
```diff
@@ -173,6 +174,47 @@
...
+            $scope.form.dto = PrimeVaccinationSchedule.get({id: id}, function() {
+                $('#primeVaccinationScheduleModal').modal('show');
+                $scope.reloadSelects();
+            });
+        };
+
+        $scope.savePrimeVaccinationSchedule = function() {
+            PrimeVaccinationSchedule.addOrUpdate($scope.form.dto,
+                function success() {
+                    $("#visitBookingDetails").trigger('reloadGrid');
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/controllers.js]: is this name correct?

File: booking-app/src/main/resources/webapp/js/services.js [hunk 24]
```diff
@@ -17,7 +13,16 @@
     });
 
     services.factory('Sites', function($resource) {
-        return $resource('../booking-app/sites', {}, {});
+        return $resource('../booking-app/sites', {}, {
+            'getClinics': { url: '../booking-app/sites/:id/clinics', method: 'GET', isArray: true },
+            'getRooms': { url: '../booking-app/sites/clinics/:id/rooms', method: 'GET', isArray: true }
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/js/services.js]: rooms are not used any more

File: booking-app/src/main/resources/webapp/partials/volunteerCardScreening.html [hunk 25]
```diff
@@ -11,7 +11,7 @@
     <table id="volunteer">
         <tr>
             <th>Booking ID</th>
-            <td id="bookingId"></td>
+            <td id="detailsId"></td>
```
Comment (pmuchowski) [inline booking-app/src/main/resources/webapp/partials/volunteerCardScreening.html]: why this is renamed?
